Index: python3.13/Modules/_hashopenssl.c
===================================================================
--- python3.13.orig/Modules/_hashopenssl.c
+++ python3.13/Modules/_hashopenssl.c
@@ -39,6 +39,10 @@
 
 #include <stdbool.h>
 
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+#include <openssl/params.h>
+#endif
+
 #ifndef OPENSSL_THREADS
 #  error "OPENSSL_THREADS is not defined, Python requires thread-safe OpenSSL"
 #endif
@@ -284,7 +288,8 @@ typedef struct {
 
 typedef struct {
     PyObject_HEAD
-    HMAC_CTX *ctx;            /* OpenSSL hmac context */
+    EVP_MAC_CTX *ctx;         /* OpenSSL hmac context */
+    EVP_MD *md;               /* OpenSSL MD */
     // Prevents undefined behavior via multiple threads entering the C API.
     bool use_mutex;
     PyMutex mutex;  /* HMAC context lock */
@@ -310,7 +315,11 @@ set_ssl_exception_from_errcode(PyObject
 
     /* ERR_ERROR_STRING(3) ensures that the messages below are ASCII */
     const char *lib = ERR_lib_error_string(errcode);
+#if (OPENSSL_VERSION_NUMBER < 0x30000000L)
     const char *func = ERR_func_error_string(errcode);
+#else
+    const char *func = "";
+#endif
     const char *reason = ERR_reason_error_string(errcode);
 
     if (lib && func) {
@@ -722,7 +731,8 @@ EVP_get_digest_size(EVPobject *self, voi
 static PyObject *
 EVP_get_name(EVPobject *self, void *closure)
 {
-    const EVP_MD *md = EVP_MD_CTX_md(self->ctx);
+    const EVP_MD *md = EVP_MD_CTX_get0_md(self->ctx);
+
     if (md == NULL) {
         notify_ssl_error_occurred();
         return NULL;
@@ -751,7 +761,7 @@ static PyObject *
 EVP_repr(EVPobject *self)
 {
     PyObject *name_obj, *repr;
-    name_obj = py_digest_name(EVP_MD_CTX_md(self->ctx));
+    name_obj = py_digest_name(EVP_MD_CTX_get0_md(self->ctx));
     if (!name_obj) {
         return NULL;
     }
@@ -1537,7 +1547,7 @@ _hashlib_hmac_singleshot_impl(PyObject *
 /*[clinic end generated code: output=82f19965d12706ac input=0a0790cc3db45c2e]*/
 {
     unsigned char md[EVP_MAX_MD_SIZE] = {0};
-    unsigned int md_len = 0;
+    size_t md_len = 0;
     unsigned char *result;
     PY_EVP_MD *evp;
 
@@ -1558,11 +1568,10 @@ _hashlib_hmac_singleshot_impl(PyObject *
     }
 
     Py_BEGIN_ALLOW_THREADS
-    result = HMAC(
-        evp,
-        (const void *)key->buf, (int)key->len,
-        (const unsigned char *)msg->buf, (size_t)msg->len,
-        md, &md_len
+    result = EVP_Q_mac(NULL, "HMAC", NULL, EVP_MD_name(evp), NULL,
+        (const void*)key->buf, key->len,
+        (const unsigned char*)msg->buf, msg->len,
+        md, sizeof(md), &md_len
     );
     Py_END_ALLOW_THREADS
     PY_EVP_MD_free(evp);
@@ -1582,11 +1591,10 @@ static int _hmac_update(HMACobject*, PyO
 static const EVP_MD *
 _hashlib_hmac_get_md(HMACobject *self)
 {
-    const EVP_MD *md = HMAC_CTX_get_md(self->ctx);
-    if (md == NULL) {
+    if (self->md == NULL) {
         raise_ssl_error(PyExc_ValueError, "missing EVP_MD for HMAC context");
     }
-    return md;
+    return self->md;
 }
 
 /*[clinic input]
@@ -1605,7 +1613,9 @@ _hashlib_hmac_new_impl(PyObject *module,
 /*[clinic end generated code: output=c20d9e4d9ed6d219 input=5f4071dcc7f34362]*/
 {
     PY_EVP_MD *digest;
-    HMAC_CTX *ctx = NULL;
+    OSSL_PARAM params[2];
+    EVP_MAC *mac = NULL;
+    EVP_MAC_CTX *ctx = NULL;
     HMACobject *self = NULL;
     int r;
 
@@ -1626,15 +1636,29 @@ _hashlib_hmac_new_impl(PyObject *module,
         return NULL;
     }
 
-    ctx = HMAC_CTX_new();
+    mac = EVP_MAC_fetch(NULL, "HMAC", NULL);
+    if (mac == NULL) {
+        _setException(PyExc_ValueError, NULL);
+       goto error;
+    }
+
+    ctx = EVP_MAC_CTX_new(mac);
     if (ctx == NULL) {
         PY_EVP_MD_free(digest);
         PyErr_NoMemory();
         goto error;
     }
 
-    r = HMAC_Init_ex(ctx, key->buf, (int)key->len, digest, NULL /* impl */);
-    PY_EVP_MD_free(digest);
+    params[0] = OSSL_PARAM_construct_utf8_string("digest",
+        (char *) EVP_MD_name(digest), 0);
+    params[1] = OSSL_PARAM_construct_end();
+
+    r = EVP_MAC_init(
+       ctx,
+       (const unsigned char*)key->buf,
+       (size_t)key->len,
+       params);
+
     if (r == 0) {
         notify_ssl_error_occurred();
         goto error;
@@ -1646,6 +1670,7 @@ _hashlib_hmac_new_impl(PyObject *module,
         goto error;
     }
 
+    self->md = digest;
     self->ctx = ctx;
     ctx = NULL;  // 'ctx' is now owned by 'self'
     HASHLIB_INIT_MUTEX(self);
@@ -1658,21 +1683,13 @@ _hashlib_hmac_new_impl(PyObject *module,
     return (PyObject *)self;
 
 error:
-    if (ctx) HMAC_CTX_free(ctx);
+    if (digest) PY_EVP_MD_free(digest);
+    if (ctx) EVP_MAC_CTX_free(ctx);
     Py_XDECREF(self);
     return NULL;
 }
 
 /* helper functions */
-static int
-locked_HMAC_CTX_copy(HMAC_CTX *new_ctx_p, HMACobject *self)
-{
-    int result;
-    ENTER_HASHLIB(self);
-    result = HMAC_CTX_copy(new_ctx_p, self->ctx);
-    LEAVE_HASHLIB(self);
-    return result;
-}
 
 /* returning 0 means that an error occurred and an exception is set */
 static unsigned int
@@ -1682,12 +1699,12 @@ _hashlib_hmac_digest_size(HMACobject *se
     if (md == NULL) {
         return 0;
     }
-    unsigned int digest_size = EVP_MD_size(md);
+    size_t digest_size = EVP_MAC_CTX_get_mac_size(self->ctx);
     assert(digest_size <= EVP_MAX_MD_SIZE);
     if (digest_size == 0) {
         raise_ssl_error(PyExc_ValueError, "invalid digest size");
     }
-    return digest_size;
+    return (unsigned int) digest_size;
 }
 
 static int
@@ -1704,15 +1721,11 @@ _hmac_update(HMACobject *self, PyObject
     if (self->use_mutex) {
         Py_BEGIN_ALLOW_THREADS
         PyMutex_Lock(&self->mutex);
-        r = HMAC_Update(self->ctx,
-                        (const unsigned char *)view.buf,
-                        (size_t)view.len);
+        r = EVP_MAC_update(self->ctx, (const unsigned char*)view.buf, view.len);
         PyMutex_Unlock(&self->mutex);
         Py_END_ALLOW_THREADS
     } else {
-        r = HMAC_Update(self->ctx,
-                        (const unsigned char *)view.buf,
-                        (size_t)view.len);
+        r = EVP_MAC_update(self->ctx, (const unsigned char*)view.buf, view.len);
     }
 
     PyBuffer_Release(&view);
@@ -1736,22 +1749,19 @@ _hashlib_HMAC_copy_impl(HMACobject *self
 {
     HMACobject *retval;
 
-    HMAC_CTX *ctx = HMAC_CTX_new();
+    EVP_MAC_CTX *ctx = EVP_MAC_CTX_dup(self->ctx);
     if (ctx == NULL) {
         return PyErr_NoMemory();
     }
-    if (!locked_HMAC_CTX_copy(ctx, self)) {
-        HMAC_CTX_free(ctx);
-        notify_ssl_error_occurred();
-        return NULL;
-    }
 
     retval = PyObject_New(HMACobject, Py_TYPE(self));
     if (retval == NULL) {
-        HMAC_CTX_free(ctx);
+        EVP_MAC_CTX_free(ctx);
         return NULL;
     }
-    retval->ctx = ctx;
+    PY_EVP_MD_up_ref(self->md);
+    retval->md = self->md;
+    retval->ctx = self->ctx;
     HASHLIB_INIT_MUTEX(retval);
 
     return (PyObject *)retval;
@@ -1761,8 +1771,12 @@ static void
 _hmac_dealloc(HMACobject *self)
 {
     PyTypeObject *tp = Py_TYPE(self);
+    if (self->md != NULL) {
+        PY_EVP_MD_free(self->md);
+        self->md = NULL;
+    }
     if (self->ctx != NULL) {
-        HMAC_CTX_free(self->ctx);
+        EVP_MAC_CTX_free(self->ctx);
         self->ctx = NULL;
     }
     PyObject_Free(self);
@@ -1807,18 +1821,14 @@ _hashlib_HMAC_update_impl(HMACobject *se
 static int
 _hmac_digest(HMACobject *self, unsigned char *buf, unsigned int len)
 {
-    HMAC_CTX *temp_ctx = HMAC_CTX_new();
+    size_t outsize = len;
+    EVP_MAC_CTX *temp_ctx = EVP_MAC_CTX_dup(self->ctx);
     if (temp_ctx == NULL) {
         (void)PyErr_NoMemory();
         return 0;
     }
-    if (!locked_HMAC_CTX_copy(temp_ctx, self)) {
-        HMAC_CTX_free(temp_ctx);
-        notify_ssl_error_occurred();
-        return 0;
-    }
-    int r = HMAC_Final(temp_ctx, buf, &len);
-    HMAC_CTX_free(temp_ctx);
+    int r = EVP_MAC_final(temp_ctx, buf, &outsize, outsize);
+    EVP_MAC_CTX_free(temp_ctx);
     if (r == 0) {
         notify_ssl_error_occurred();
         return 0;
@@ -1882,8 +1892,11 @@ _hashlib_hmac_get_digest_size(HMACobject
 static PyObject *
 _hashlib_hmac_get_block_size(HMACobject *self, void *closure)
 {
-    const EVP_MD *md = _hashlib_hmac_get_md(self);
-    return md == NULL ? NULL : PyLong_FromLong(EVP_MD_block_size(md));
+    if (self->ctx == NULL) {
+        return NULL;
+    }
+
+    return PyLong_FromLong(EVP_MAC_CTX_get_block_size(self->ctx));
 }
 
 static PyObject *
