Index: python3.13/Modules/_ssl.c
===================================================================
--- python3.13.orig/Modules/_ssl.c
+++ python3.13/Modules/_ssl.c
@@ -131,6 +131,22 @@ static void _PySSLFixErrno(void) {
 #error Unsupported OpenSSL version
 #endif
 
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+/* Disable version methods when building for OpenSSL 3 */
+#ifndef OPENSSL_NO_SSL3_METHOD
+#define OPENSSL_NO_SSL3_METHOD 1
+#endif
+#ifndef OPENSSL_NO_TLS1_METHOD
+#define OPENSSL_NO_TLS1_METHOD 1
+#endif
+#ifndef OPENSSL_NO_TLS1_1_METHOD
+#define OPENSSL_NO_TLS1_1_METHOD 1
+#endif
+#ifndef OPENSSL_NO_TLS1_2_METHOD
+#define OPENSSL_NO_TLS1_2_METHOD 1
+#endif
+#endif
+
 /* OpenSSL API 1.1.0+ does not include version methods */
 #ifndef OPENSSL_NO_SSL3_METHOD
 extern const SSL_METHOD *SSLv3_method(void);
@@ -1852,7 +1868,7 @@ _ssl__SSLSocket_getpeercert_impl(PySSLSo
                         "handshake not done yet");
         return NULL;
     }
-    peer_cert = SSL_get_peer_certificate(self->ssl);
+    peer_cert = SSL_get1_peer_certificate(self->ssl);
     if (peer_cert == NULL)
         Py_RETURN_NONE;
 
@@ -4369,7 +4385,8 @@ _ssl__SSLContext_load_dh_params_impl(PyS
 /*[clinic end generated code: output=dd74b3c524dd2723 input=832769a0734b8c4d]*/
 {
     FILE *f;
-    DH *dh;
+    BIO *b;
+    EVP_PKEY *pkey;
 
 #if defined(MS_WINDOWS) && defined(_DEBUG)
     PyErr_SetString(PyExc_NotImplementedError,
@@ -4381,12 +4398,19 @@ _ssl__SSLContext_load_dh_params_impl(PyS
     if (f == NULL)
         return NULL;
 
+    b = BIO_new_fp(f, 0);
+    if (b == NULL) {
+        fclose(f);
+        return NULL;
+    }
+
     errno = 0;
     PySSL_BEGIN_ALLOW_THREADS
-    dh = PEM_read_DHparams(f, NULL, NULL, NULL);
+    pkey = PEM_read_bio_Parameters(b, NULL);
+    BIO_free(b);
     fclose(f);
     PySSL_END_ALLOW_THREADS
-    if (dh == NULL) {
+    if (pkey == NULL) {
         if (errno != 0) {
             PyErr_SetFromErrnoWithFilenameObject(PyExc_OSError, filepath);
             ERR_clear_error();
@@ -4396,11 +4420,17 @@ _ssl__SSLContext_load_dh_params_impl(PyS
         }
         return NULL;
     }
-    if (!SSL_CTX_set_tmp_dh(self->ctx, dh)) {
-        DH_free(dh);
+#if (OPENSSL_VERSION_NUMBER < 0x30000000L)
+    if (!SSL_CTX_set_tmp_dh(self->ctx, EVP_PKEY_get0_DH(pkey))) {
+        EVP_PKEY_free(pkey);
         return _setSSLError(get_state_ctx(self), NULL, 0, __FILE__, __LINE__);
     }
-    DH_free(dh);
+#else
+    if (!SSL_CTX_set0_tmp_dh_pkey(self->ctx, pkey)) {
+        EVP_PKEY_free(pkey);
+        return _setSSLError(get_state_ctx(self), NULL, 0, __FILE__, __LINE__);
+    }
+#endif
     Py_RETURN_NONE;
 }
 
