--- a/Lib/test/test_hashlib.py
+++ b/Lib/test/test_hashlib.py
@@ -151,7 +151,8 @@
         if _hashlib:
             # These algorithms should always be present when this module
             # is compiled.  If not, something was compiled wrong.
-            self.assertTrue(hasattr(_hashlib, 'openssl_md5'))
+            # Pexip: We remove the openssl md5 implementation
+            # self.assertTrue(hasattr(_hashlib, 'openssl_md5'))
             self.assertTrue(hasattr(_hashlib, 'openssl_sha1'))
             for algorithm, constructors in self.constructors_to_test.items():
                 constructor = getattr(_hashlib, 'openssl_'+algorithm, None)
--- a/Modules/_hashopenssl.c
+++ b/Modules/_hashopenssl.c
@@ -100,7 +100,6 @@
 } py_hashentry_t;
 
 // Fundamental to TLS, assumed always present in any libcrypto:
-#define Py_hash_md5 "md5"
 #define Py_hash_sha1 "sha1"
 #define Py_hash_sha224 "sha224"
 #define Py_hash_sha256 "sha256"
@@ -143,8 +142,6 @@
     {py_name, py_alias, ossl_name, ossl_nid, 0, NULL, NULL}
 
 static const py_hashentry_t py_hashes[] = {
-    /* md5 */
-    PY_HASH_ENTRY(Py_hash_md5, "MD5", SN_md5, NID_md5),
     /* sha1 */
     PY_HASH_ENTRY(Py_hash_sha1, "SHA1", SN_sha1, NID_sha1),
     /* sha2 family */
@@ -1073,27 +1070,6 @@
 
 
 /*[clinic input]
-_hashlib.openssl_md5
-
-    data: object(c_default="NULL") = b''
-    *
-    usedforsecurity: bool = True
-    string: object(c_default="NULL") = None
-
-Returns a md5 hash object; optionally initialized with a string
-
-[clinic start generated code]*/
-
-static PyObject *
-_hashlib_openssl_md5_impl(PyObject *module, PyObject *data,
-                          int usedforsecurity, PyObject *string)
-/*[clinic end generated code: output=ca8cf184d90f7432 input=e7c0adbd6a867db1]*/
-{
-    CALL_HASHLIB_NEW(module, Py_hash_md5, data, string, usedforsecurity);
-}
-
-
-/*[clinic input]
 _hashlib.openssl_sha1
 
     data: object(c_default="NULL") = b''
@@ -2205,7 +2181,6 @@
     _HASHLIB_COMPARE_DIGEST_METHODDEF
     _HASHLIB_HMAC_SINGLESHOT_METHODDEF
     _HASHLIB_HMAC_NEW_METHODDEF
-    _HASHLIB_OPENSSL_MD5_METHODDEF
     _HASHLIB_OPENSSL_SHA1_METHODDEF
     _HASHLIB_OPENSSL_SHA224_METHODDEF
     _HASHLIB_OPENSSL_SHA256_METHODDEF
